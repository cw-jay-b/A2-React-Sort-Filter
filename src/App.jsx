import React, { useState, useEffect } from 'react';
import axios from 'axios';
import SortFilterSection from './components/SortFilterSection';
import ProductCard from './components/ProductCard';
import NoDataCard from './components/NoDataCard';
import './styles/App.css';


import { stocks } from "./dummy/data/cars.json";

const App = () => {
    const [totalCount, setTotalCount] =useState(0);
    const fuelTypes = ['Petrol', 'Diesel', 'CNG', 'LPG', 'Electric', 'Hybrid'];
    const [products, setProducts] = useState([]);
    const [filteredProducts, setFilteredProducts] = useState([]);
    const [selectedFuel, setSelectedFuel] = useState([]);
    const [priceRange, setPriceRange] = useState({ min: 0, max: Number.MAX_SAFE_INTEGER });
    const [sortOrder, setSortOrder] = useState('asc');
    useEffect(() => {
        const fetchData = async () => {
            // *************************** DUMMY DATA TO BE REMOVED LATER ***************************

            // setProducts(stocks)

            // **************************************************************************************

            // console.log(productsResponse.data.stocks);


            // console.log(productsResponse);
            // console.log(productsResponse.data);
            // console.log(productsResponse.data.stocks);
            try {
                const [productsResponse, categoriesResponse] = await Promise.all([
                    // https://stg.carwale.com/api/stocks?fuel=1+2
                    // axios.get('https://dummyjson.com/products'),
                    // axios.get('https://dummyjson.com/c/2ddb-0dd0-4899-aef2'),
                    // axios.get('https://stg.carwale.com/api/stocks?fuel=1+2+3+4+5+6&budget=0-'),
                    axios.get('/api/stocks'),
                    axios.get('https://dummyjson.com/products/categories')
                ]);
                setTotalCount(productsResponse.data.totalCount);
                setProducts(productsResponse.data.stocks);
                setFilteredProducts(productsResponse.data.stocks);
            } catch (error) {
                console.error('Error fetching data:', error);
            }
        };
        fetchData();
    }, []);

    const applyFilters = () => {
        let filtered = [...products];

        // Create a mapping of selected fuel numbers to fuel names
        const selectedFuelNames = selectedFuel.map(number => fuelTypes[number - 1]);

        if (selectedFuelNames.length > 0) {
            filtered = filtered.filter(product => selectedFuelNames.includes(product.fuel));
        }

        // Filter by price range
        filtered = filtered.filter(
            product => parseInt(product.priceNumeric) >= priceRange.min && parseInt(product.priceNumeric) <= priceRange.max
        );

        filtered.sort((a, b) => {
            if (sortOrder === 'asc') {
                return (a.priceNumeric - b.priceNumeric); // Ascending price
            } else {
                return (b.priceNumeric - a.priceNumeric); // Descending price
            }
        });
    
        setFilteredProducts(filtered);
    };

    useEffect(() => {
        applyFilters();
    }, [selectedFuel, priceRange, sortOrder, products]);

    return (
       <>
       <h1>{totalCount} Used Cars In India</h1>
       
       <div className="content-wrapper" >
        <section className="left" >
            <SortFilterSection
                selectedFuel={selectedFuel}
                setSelectedFuel={setSelectedFuel}
                priceRange={priceRange}
                setPriceRange={setPriceRange}
                sortOrder={sortOrder}
                setSortOrder={setSortOrder}
            />
            <div className="results-count" >Showing {filteredProducts.length} {filteredProducts.length>1?"Cars":"Car"} </div>  
        </section>
        <section className='right' >
            <div className='sort-wrapper'>
            <h5 style={{padding: "0.5rem 0", fontWeight: "normal"}}>Sort by Price</h5>
            <select className='sort-input-box' value={sortOrder} onChange={(e) => setSortOrder(e.target.value)}>
                <option value="asc">Ascending</option>
                <option value="desc">Descending</option>
            </select>
            </div>
            <div className="products-wrapper">
            {
                filteredProducts.length > 0 ? (
                    filteredProducts.map(product => (
                        <ProductCard key={product.profileId} {...product} />
                    ))
                ) : (
                    <NoDataCard/>
                )
            }
            </div>
        </section>
       </div>
       </> 
    );
}

export default App;