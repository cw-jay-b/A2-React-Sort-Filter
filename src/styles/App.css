main {
  display: flex;
  flex-direction: column;
  /* height: 100%; */
  /* overflow: hidden; */
  align-items: flex-end;
}

.filter-section {
  border: 1px solid lightgray;
  border-radius: 0.5rem;
  background-color: #fff;
  padding: 20px;
  border-right: 1px solid #ddd;
}

.content-wrapper{
  display: flex;
  column-gap: 1rem;
}

.sort-section {
  display: flex;
  position: sticky;
  top: 0;
  background-color: #f9f9f9;
  padding: 10px;
  z-index: 10;
  border-bottom: 1px solid #ddd;
  justify-content: flex-end;
  column-gap: 0.5rem;
}

.results-count{
  margin-top: 1rem;
  font-size: 1.25rem;
}