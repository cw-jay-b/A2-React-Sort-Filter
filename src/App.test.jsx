import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import '@testing-library/jest-dom';
import App from './App';
import axios from 'axios';

// Mock axios
vi.mock('axios');

describe('App Component', () => {

  it('changes sorting option from ascending to descending', async () => {
    axios.get.mockResolvedValue({
      data: {
        totalCount: 100,
        stocks: [
          {profileId: '1', stockImages:[],priceNumeric: '150000', fuel: 'Petrol' },
          {profileId: '2', stockImages:[],priceNumeric: '250000', fuel: 'Diesel' }
        ]
      }
    });

    render(<App />);

    expect(screen.getByRole('combobox')).toHaveTextContent("Ascending");

    fireEvent.change(screen.getByRole('combobox'), {
      target: { value: 'desc' }
    });

    expect(screen.getByRole('combobox')).toHaveTextContent("Descending");
  });
});
