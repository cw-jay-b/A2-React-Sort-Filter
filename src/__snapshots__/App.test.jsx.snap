// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Component > matches snapshot with total count 1`] = `
<DocumentFragment>
  <h1
    style="margin-bottom: 1rem;"
  >
    0 Used Cars In India
  </h1>
  <div
    class="content-wrapper"
  >
    <section
      class="left"
    >
      <aside
        class="filter-section"
        style="position: sticky; top: 1rem;"
      >
        <div
          style="margin-bottom: 1rem; display: flex; justify-content: space-between;"
        >
          <span>
            <img
              alt="filter_icon"
              src="/src/assets/icons8-filter-16.png"
              style="margin-right: 0.5rem;"
            />
            <span
              style="font-size: 1.25rem; font-weight: bold;"
            >
              Filters
            </span>
          </span>
          <button
            style="color: rgb(9, 121, 182); background: none; outline: none; cursor: pointer;"
          >
            Clear All
          </button>
        </div>
        <div
          class="filter-group"
        >
          <h4>
            Fuel
          </h4>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="1"
            />
            Petrol
          </label>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="2"
            />
            Diesel
          </label>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="3"
            />
            CNG
          </label>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="4"
            />
            LPG
          </label>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="5"
            />
            Electric
          </label>
          <label>
            <input
              style="margin: 0px 0.5rem;"
              type="checkbox"
              value="6"
            />
            Hybrid
          </label>
        </div>
        <div
          class="filter-group"
          id="min-max-filter"
          style="margin-bottom: 0px;"
        >
          <h4>
            Price Range
          </h4>
          <div
            style="display: flex; align-items: center;"
          >
            <span
              class="min-max-box"
              id="min-box"
            >
              <input
                id="min-price-input"
                min="0"
                name="min"
                placeholder="1"
                type="number"
              />
              Lakhs
            </span>
             
            <span>
              -
            </span>
            <span
              class="min-max-box"
              id="max-box"
            >
              <input
                id="max-price-input"
                min="0"
                name="max"
                placeholder="5"
                type="number"
              />
              Lakhs
            </span>
          </div>
        </div>
        <div
          class="results-count"
        >
          Showing 0 Car 
        </div>
      </aside>
    </section>
    <section
      class="right"
    >
      <div
        class="sort-wrapper"
      >
        <h5
          style="padding: 0.5rem 0px; font-weight: normal;"
        >
          Sort By:
        </h5>
        <select
          class="sort-input-box"
        >
          <option
            value="asc"
          >
            Price - Ascending
          </option>
          <option
            value="desc"
          >
            Price - Descending
          </option>
        </select>
      </div>
      <div
        class="products-wrapper"
      >
        <div
          class="no-data-card"
        >
          No vehicles found...
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
